### PR TITLE
Add data fingerprint file

### PR DESCRIPTION
**Pull request by Manishearth**
This adds the ability to generate "fingerprint" files that have data hashes for JSON and postcard data for each key/locale, sorted alphabetically. This can be used to ensure data format stability by tracking changes to our data in these two formats, while being able to tell the difference between "key added" and "data changed".

This does include a JSON fingerprint even though JSON stability can be easily noticed by looking at the diffstat on the PR. It's more useful for Postcard, where all changes to postcard data occur opaquely inside a single blob.

Fingerprinting is automatically run when regenerating testdata. Reviewers to testdata should be careful in ensuring that the change in a hash for a key is expected.

Fixes https://github.com/unicode-org/icu4x/issues/538

cc @robertbastian

### CODE REVIEW

File: provider/datagen/src/bin/fingerprint-data.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                       or file containing data generated by icu4x-datagen --format=blob.",
+                ),
+        )
+        .arg(
+            Arg::with_name("FORMAT")
+                .long("format")
+                .takes_value(true)
+                .possible_value("dir")
+                .possible_value("blob")
+                .help("Output to a directory on the filesystem or a single blob.")
```
Comment (robertbastian) [inline provider/datagen/src/bin/fingerprint-data.rs:41]: Nit: update the doc. You can even remove the arg, we can detect whether a path is a dir or a file.

**Review by robertbastian — APPROVED** [review]
Looks good but...

... I'm not convinced we actually need this. We already have the functionality to build a Postcard tree, this just pulls in more dependencies and adds more code that we have to maintain. Checking in the actual postcard data is easy and more useful imho as it allows us to see if a file has grown or shrunk, or even moved (which might not be immediately obvious with the hash file). The postcard tree is 4MB at the moment which isn't much, and if we wanted we can also exclude it from `cargo publish`.

**Comment by Manishearth** [comment]
@robertbastian We currently don't check in postcard FS data as testdata, and I don't think we _should_, i'd like testdata to be the data we use in our `cargo test` tests.

I find that having a pair of files that contains the entire data change history will be quite valuable, especially for things like release steps since we can then do a review of the fingerprint diff between releases as a pre-release verification step.


As far as deps: this only adds one dep, `pathdiff`, and it's a small crate I maintain that contains a function that was removed from Rust's stdlib. It's pretty widely used. `walkdir` and `sha2` are already in our transitive deptree. Overall we're not too picky about datagen deps anyway.

A downside of this PR is that it hashes the _data_ as opposed to the schema, so CLDR updates will cause data changes. I have some thoughts on this which I'll post to https://github.com/unicode-org/icu4x/issues/538.

File: provider/blob/src/blob_data_provider.rs [hunk 2]
```diff
@@ -69,6 +69,12 @@ impl BlobDataProvider {
             })?,
         })
     }
+
+    #[cfg(feature = "export")]
+    #[doc(hidden)] // See #1771, we don't want this to be a publicly visible API
+    pub fn get_map(&self) -> &ZeroMap2dBorrowed<ResourceKeyHash, str, [u8]> {
```
Comment (sffc) [inline provider/blob/src/blob_data_provider.rs:75]: Thought: You know how I feel about making BlobDataProvider be iterable. At least this is hidden and used only by test code, but I have a slight preference for holding this by about a week, at which point ResourceOptions will be much-improved and we can use standard APIs for this.

File: provider/datagen/Cargo.toml [hunk 3]
```diff
@@ -75,6 +75,9 @@ zerovec = { version = "0.6", path = "../../utils/zerovec", features = ["serde_se
 reqwest = { version = "0.11", features = ["stream"], optional = true }
 tokio = { version = "1.5", features = ["rt-multi-thread", "macros", "fs"], optional = true }
 futures = { version = "0.3", optional = true }
+walkdir = "2.3.2"
+sha2 = "0.10.2"
+pathdiff = "0.2.1"
```
Comment (sffc) [inline provider/datagen/Cargo.toml:80]: Nit: If these are not "dependencies for the download-testdata feature" then please put them up above in alphabetical order

File: provider/blob/src/blob_data_provider.rs (hunk 2 above)
Comment (Manishearth) [reply provider/blob/src/blob_data_provider.rs:75]: Sounds good, though personally I'd prefer to land this and then use this code as the first client to test out iterable resource options

Comment (sffc) [reply provider/blob/src/blob_data_provider.rs:75]: ok.

**Review by sffc — APPROVED** [review]

**Pull request merged**